fms_extras/utils/cache/__init__.py

```diff
@@ -0,0 +1,245 @@
+    def is_filled(self) -> bool:
+        """
+        Denotes whether this cache data layer is in post-fill stage
+
+        Returns
+        -------
+        bool
+            True if cache data layer is currently in the post-fill stage, otherwise False and cache data layer is being
+            pre-filled
+        """
+        pass
+
+
+@dataclasses.dataclass
+class CacheData(metaclass=abc.ABCMeta):
+    """
+    Dataclass responsible for holding raw cache data.
+
+    Attributes
+    ----------
+    data: List[Tuple[torch.Tensor, torch.Tensor]]
+        data is represented as a List of tuples of Tensors where each index in the list corresponds to a layer in the
+        cache, and each position in the tuple corresponds to the key and value cache block for that layer
+    max_sequence_length: int
+        max sequence length of all sequences stored in this cache data
+    """
+
+    data: List[Tuple[torch.Tensor, torch.Tensor]]
+    max_sequence_length: int
+
+    @abc.abstractmethod
+    def get_layer(self, layer_index: int) -> CacheDataLayer:
+        """
+        Get a single layer of the cache data
+
+        Parameters
+        ----------
+        layer_index: int
+            index of layer
+
+        Returns
+        -------
+        CacheDataLayer
+            a single layer of the cache data as a dataclass
+        """
+        pass
+
+    @abc.abstractmethod
+    def is_filled(self) -> bool:
+        """
+        Determines if the cache has been filled with the prompt, or is completely empty for this piece of cache data
+
+        Returns
+        -------
+        bool
+            True if the keys and values for the prompt have been set, otherwise False.
+        """
+        pass
+
+
+@dataclasses.dataclass
+class CacheDataWithMetadata(CacheData):
+    """A special form of CacheData that includes some simple metadata associated with it
+
+    Attributes
+    ----------
+    data: List[Tuple[torch.Tensor, torch.Tensor]]
+        data is represented as a List of tuples of Tensors where each index in the list corresponds to a layer in the
+        cache, and each position in the tuple corresponds to the key and value cache block for that layer
+    max_sequence_length: int
+        max sequence length of all sequences corresponding to the sequence ids in this cache data
+    sequence_ids: List[int]
+        the integer ids associated with each sequence, these will correspond by index with the input ids passed to the
+        model
+    context_lengths: torch.Tensor, optional
+        a 1d tensor corresponding to the length of each sequence in the batch denoted by the sequence ids. If None,
+        cache data is in pre-fill
+    """
+
+    data: List[Tuple[torch.Tensor, torch.Tensor]]
+    max_sequence_length: int
+    sequence_ids: List[int]
+    context_lengths: Optional[torch.Tensor]
+
+    def compute_position_ids(self, num_tokens_per_sequence: List[int]) -> torch.Tensor:
+        position_ids_list = compute_position_ids(
```

Comment:
this is a bit confusing, looks as if it's recursive but isn't

Reply:
Removed the extra utility function as it is not used.